tokio/src/io/driver/scheduled_io.rs

```diff
@@ -1,28 +1,106 @@
-use crate::loom::future::AtomicWaker;
+use super::{platform, Direction, ReadyEvent, Tick};
 use crate::loom::sync::atomic::AtomicUsize;
+use crate::loom::sync::Mutex;
+use crate::util::bit;
 use crate::util::slab::Entry;
 
 use std::sync::atomic::Ordering::{AcqRel, Acquire, Release};
+use std::task::{Context, Poll, Waker};
+
+cfg_io_readiness! {
+    use crate::util::linked_list::{self, LinkedList};
+
+    use std::cell::UnsafeCell;
+    use std::future::Future;
+    use std::marker::PhantomPinned;
+    use std::pin::Pin;
+    use std::ptr::NonNull;
+}
 
 /// Stored in the I/O driver resource slab.
 #[derive(Debug)]
 pub(crate) struct ScheduledIo {
     /// Packs the resource's readiness with the resource's generation.
     readiness: AtomicUsize,
 
-    /// Task waiting on read readiness
-    pub(crate) reader: AtomicWaker,
+    waiters: Mutex<Waiters>,
+}
+
+#[cfg(feature = "io-readiness")]
+type WaitList = LinkedList<Waiter, <Waiter as linked_list::Link>::Target>;
```

Comment:
this makes me wonder again if there is potentially an internal waitlist primitive to factor out that could be common between `Notify`, `Semaphore`, and the IO driver. But, that seems like something to think about in follow-ups.